lib/bonus.stk

```diff
@@ -1124,6 +1127,109 @@ doc>
       (begin ,@body
              (-> #:top)))))
 
+#|
+<doc EXT 1+ 1-
+ * (1+ x)
+ * (1- x)
+ *
+ * There procedures return |x| plus one and |x| minus one, repsectively.
+ * Being procedures, they do not change the value of the variable |x|.
+ *
+ * @lisp
+ * (define x 10)
+ * (1+ x)          => 11
+ * x               => 10
+ * (1- x)          => 9
+ * x               => 10
+ * @end lisp
+doc>
+|#
+(define (1+ x) (+ x 1))
+(define (1- x) (- x 1))
+
+
+#|
+<doc EXT inc! dec!
+ * (inc! x)
+ * (dec! x)
+ *
+ * There macros change the value of |x| to |x|+1 and |x|-1, respectively.
+ *
+ * @lisp
+ * (define x 10)
+ * (inc! x)
+ * x               => 11
+ * (dec! x)
+ * x               => 10
+ * @end lisp
+doc>
+|#
+(define-macro (inc! x)
+  `(set! ,x (+ ,x 1)))
```

Comment:
This will cause trouble if you have generalized `set!` in the future because `x` is evaluted twice. Unfortunately, dealing with generalized `set!` efficiently is hard; see "setf expansions" in Common Lisp.

Reply:
Oh, you're right! In fact, STklos *does* have generalized `set!`, and I forgot about that. I'll remove this part for now.

Reply:
`push!` and `pop!` evaluate the list twice, too.

Reply:
Yes - the expansion should be more careful. I'll work more on this later...

Reply:
For example, Gauche expects either a symbol or an s-expression as the "stack" argument, and expands it as this:
```scheme
(macroexpand '(push! (list 1 2 3) 10))
  =>
(let
 ((getter.0 f) 
  (#:G347 a) 
  (#:G348 b) 
  (#:G349 c))
 ((setter getter.0) #:G347 #:G348 #:G349
  (cons 10 (getter.0 #:G347 #:G348 #:G349))))
```
which is very similar to what Clisp, for example, does:
```lisp
(macroexpand '(push 10 (f a b c)))
  =>
(LET* ((#:G2953 10)) 
  (LET* ((#:TEMP-2952 A)
         (#:TEMP-2951 B)
         (#:TEMP-2950 C)) 
    (LET* ((#:G2954 (F #:TEMP-2952 #:TEMP-2951 #:TEMP-2950)))
      (LET* ((#:NEW-2949 (CONS #:G2953 #:G2954)))
        (FUNCALL #'(SETF F) 
                 #:NEW-2949
                 #:TEMP-2952 
                 #:TEMP-2951
                 #:TEMP-2950)))))
```
And this is ECL (almost the same as that of Clisp):
```lisp
(LET* ((#:G1 A) 
       (#:G2 B)
       (#:G3 C)
       (#:G4 (CONS 10 (F #:G1 #:G2 #:G3))))
  (DECLARE (:READ-ONLY #:G1 #:G2 #:G3))
  (FUNCALL #'(SETF F) #:G4 #:G1 #:G2 #:G3))
```

When the stack is a symbol, they all revert to the simple expansion

```lisp
(macroexpand '(push 10 x)) => (SETQ X (CONS 10 X))
```

Reply:
Yes. See [CLHS section 5.1: Generalized Reference](http://clhs.lisp.se/Body/05_a.htm). That's where the temp variables come from.

Reply:
In particular [5.1.1.2 Setf Expansions](http://clhs.lisp.se/Body/05_aab.htm).